charms/knative-serving/src/charm_template.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+from typing import List, Union, Optional
+
+import jinja2
+from jinja2 import Environment, FileSystemLoader
+from lightkube import Client, codecs
+from lightkube.core.exceptions import ApiError
+from lightkube.core.resource import NamespacedResource, GlobalResource
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, WaitingStatus, BlockedStatus, MaintenanceStatus
+
+from charm_helpers.exceptions import ErrorWithStatus, LeadershipError
+from lightkube_helpers import check_resources, get_first_worst_error
+
+
+class ExtendedCharmBase(CharmBase):
+    """An extended base class for Charms that includes some conveniences and basic features"""
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.log = logging.getLogger(__name__)
+        self.name = self.model.app.name
+        self.model_name = self.model.name
+
+        self.framework.observe(self.on.config_changed, self.on_config_changed)
+        self.framework.observe(self.on.install, self.on_install)
+        self.framework.observe(self.on.leader_elected, self.on_leader_elected)
+        self.framework.observe(self.on.remove, self.on_remove)
+        self.framework.observe(self.on.upgrade_charm, self.on_upgrade_charm)
+        self.framework.observe(self.on.update_status, self.on_update_status)
```

Comment:
I'm not sure if handling charm events 2 layers above the actual charm class might be a good idea, this might cause a lot of unexpected behaviors as the charm developer has no context on what events are being observed.

Reply:
Is this a problem with the pattern, or a documentation/training issue?  Agreed that we don't want unexpected behaviours, but this sounds like something that can be addressed by a few paragraphs in a docstring no?

Reply:
To start, I've added one in the KubernetesManifestCharmBase, can you take a look and give me your thoughts?  Some of this advice would also apply to the higher level base charm too but no need to debate it in two places at once